sync-core/src/test/java/com/cloudant/sync/query/AbstractQueryExecutorTestBase.java

```diff
@@ -0,0 +1,280 @@
+//  Copyright (c) 2014 Cloudant. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+//  except in compliance with the License. You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software distributed under the
+//  License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+//  either express or implied. See the License for the specific language governing permissions
+//  and limitations under the License.
+
+package com.cloudant.sync.query;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.contains;
+import static org.hamcrest.Matchers.containsInAnyOrder;
+import static org.hamcrest.Matchers.is;
+import static org.hamcrest.Matchers.notNullValue;
+import static org.hamcrest.Matchers.nullValue;
+
+import com.cloudant.sync.datastore.DocumentBodyFactory;
+import com.cloudant.sync.datastore.DocumentRevision;
+import com.cloudant.sync.datastore.MutableDocumentRevision;
+
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.io.IOException;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ *  The aim is to make sure that the post hoc matcher class behaves the
+ *  same as the SQL query engine.
+ *
+ *  To accomplish this, we test the entire test execution pipeline contained
+ *  within this class using {@link com.cloudant.sync.query.QueryExecutorTest}
+ *  which exercises the SQL query engine matching functionality.  We then test
+ *  the same test execution pipeline contained within this class using
+ *  {@link com.cloudant.sync.query.QueryExecutorUnindexedTest} which in turn
+ *  exercises the post hoc matcher matching functionality.
+ */
+public abstract class AbstractQueryExecutorTestBase extends AbstractIndexTestBase {
+
+    @Override
+    public void setUp() throws SQLException {
```

Comment:
I changed the processing of these tests to now use mock IndexManager and QueryExecutor subclasses to force the tests to go in two directions.  Same set of tests are executed once against the SQL engine and once against the post hoc matcher.  Done in 3f1b4c2.
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-1a5092ac6698384ba572368c278436fdL47
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-0eaf4604dd4e002edfe0d41fdea2c7e3R20
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-162b0c09f735c14ea05abac3dd234d9eR27
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-c4ac637a25d1bf100c90b6b658737b56R20
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-253a3500c1d06e06260c5b09832b8986R21
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-c8d8ef5f4948719512f62057f5359599L29
https://github.com/cloudant/sync-android/commit/3f1b4c2992cab575b6d999ec536957bb10d04f57#diff-162fe9eab99c2b46b7150404a5da366dL28